Assets/Scripts/SS3D/Systems/Substances/SubstanceContainerActor.cs

```diff
@@ -0,0 +1,618 @@
+            };
+        }
+
+        /// <summary>
+        /// Processes substances in a container
+        /// </summary>
+        public static void ProcessContainer(SubstanceContainerActor container)
+        {
+            SubstanceContainer.ProcessContainer(container.substanceContainer);
+            container.MarkDirty();
+        }
+
+        #endregion
+
+        #region SubstanceContainer
+        public class SubstanceContainer
+        {
+            /// <summary>
+            /// do not modify directly, use the property setter Volume instead.
+            /// </summary>
+            private float _volume;
+
+            private float _currentVolume;
+
+            /// <summary>
+            /// A list of all substances in this container
+            /// </summary>
+            private List<SubstanceEntry> _substances;
+
+            public ReadOnlyCollection<SubstanceEntry> Substances => _substances.AsReadOnly();
+
+            private SubstanceContainerActor _actor;
+
+            /// <summary>
+            /// Is the container locked?
+            /// </summary>
+            private bool _locked;
+
+            /// <summary>
+            /// The temperature of the container
+            /// </summary>
+            private float _temperature;
+
+            /// <summary>
+            /// The total number of moles
+            /// </summary>
+            public float TotalMoles => Substances.Sum(x => x.Moles);
+
+            /// <summary>
+            /// The remaining volume in milliliters that fit in this container
+            /// </summary>
+            public float RemainingVolume => _volume - _currentVolume;
+
+            public float Temperature => _temperature;
+
+
+            /// <summary>
+            /// The capacity of this container in milliliters
+            /// </summary>
+            public float Volume => _volume;
+
+            public bool IsEmpty => _currentVolume == 0f;
+
+            public bool CanTransfer => !_locked;
+
+            /// <summary>
+            /// Multiplier to convert moles in this container to volume
+            /// </summary>
+            public float MolesToVolume
+            {
+                get
+                {
+                    float val = 0;
+                    float total = TotalMoles;
+                    foreach (var entry in Substances)
+                    {
+                        val += entry.Substance.MillilitersPerMole * (entry.Moles / total);
+                    }
+                    return val;
+                }
+            }
+
+
+            /// <summary>
+            /// The filled volume in ml
+            /// </summary>
+            public float CurrentVolume
+            {
+                get => _currentVolume;
+                set
+                {
+                    _currentVolume = value;
+                    if (_actor != null)
+                    {
+                        _actor.currentVolume = value;
+                    }
```

Comment:
I'd move this to a Setter function, its frowned upon to make big setters with logic in them.